src/ReservoirRouter.sol

```diff
@@ -0,0 +1,305 @@
+        rPair = factory.getPair(aTokenA, aTokenB, aCurveId);
+
+        if (rPair == address(0)) {
+            rPair = factory.createPair(aTokenA, aTokenB, aCurveId);
+        }
+
+        (uint256 lReserveA, uint256 lReserveB) = ReservoirLibrary.getReserves(address(factory), aTokenA, aTokenB, aCurveId);
+        if (lReserveA == 0 && lReserveB == 0) {
+            (rAmountA, rAmountB) = (aAmountADesired, aAmountBDesired);
+        }
+        else {
+            uint lAmountBOptimal = ReservoirLibrary.quote(aAmountADesired, lReserveA, lReserveB);
+            if (lAmountBOptimal <= aAmountBDesired) {
+                require(lAmountBOptimal >= aAmountBMin, "RR: INSUFFICIENT_B_AMOUNT");
+                (rAmountA, rAmountB) = (aAmountADesired, lAmountBOptimal);
+            } else {
+                uint lAmountAOptimal = ReservoirLibrary.quote(aAmountBDesired, lReserveB, lReserveA);
+                assert(lAmountAOptimal <= aAmountADesired);
+                require(lAmountAOptimal >= aAmountAMin, "RR: INSUFFICIENT_A_AMOUNT");
+                (rAmountA, rAmountB) = (lAmountAOptimal, aAmountBDesired);
+            }
+        }
+    }
+
+    function addLiquidity(
+        address aTokenA,
+        address aTokenB,
+        uint256 aCurveId,
+        uint aAmountADesired,
+        uint aAmountBDesired,
+        uint aAmountAMin,
+        uint aAmountBMin,
+        address aTo
+    ) external payable returns (uint256 rAmountA, uint256 rAmountB, uint256 rLiq) {
+        address lPair;
+        (rAmountA, rAmountB, lPair) = _addLiquidity(aTokenA, aTokenB, aCurveId, aAmountADesired, aAmountBDesired, aAmountAMin, aAmountBMin);
+
+        _pay(aTokenA, msg.sender, lPair, rAmountA);
+        _pay(aTokenB, msg.sender, lPair, rAmountB);
+
+        rLiq = IReservoirPair(lPair).mint(aTo);
+    }
+
+    function removeLiquidity(
+        address aTokenA,
+        address aTokenB,
+        uint256 aCurveId,
+        uint256 aLiq,
+        uint256 aAmountAMin,
+        uint256 aAmountBMin,
+        address aTo
+    ) external payable returns (uint256 rAmountA, uint256 rAmountB) {
+        require(aTo != address(0), "RR: TO_ZERO_ADDRESS");
+        address lPair = ReservoirLibrary.pairFor(address(factory), aTokenA, aTokenB, aCurveId);
+        IReservoirPair(lPair).transferFrom(msg.sender, lPair, aLiq);
+        (uint lAmount0, uint lAmount1) = IReservoirPair(lPair).burn(aTo);
+
+        (address lToken0,) = ReservoirLibrary.sortTokens(aTokenA, aTokenB);
+        (rAmountA, rAmountB) = aTokenA == lToken0 ? (lAmount0, lAmount1) : (lAmount1, lAmount0);
+
+        require(rAmountA >= aAmountAMin, "RR: INSUFFICIENT_A_AMOUNT");
+        require(rAmountB >= aAmountBMin, "RR: INSUFFICIENT_B_AMOUNT");
+    }
+
+    /// @dev requires the initial amount to have already been sent to the first pair
+    function _swapExactForVariable(uint256 aAmountIn, address[] memory aPath, uint256[] memory aCurveIds, address aTo)
+        internal returns (uint256 rFinalAmount) {
```

Comment:
Needs to be `internal` or can be `private`?

Reply:
in this case, private. Fixed for both instances